app/slashcommands/command_invite.go

```diff
@@ -38,58 +37,117 @@ func (*InviteProvider) GetCommand(a *app.App, T i18n.TranslateFunc) *model.Comma
-	if len(splitMessage) > 1 && splitMessage[1] != "" {
-		targetChannelName := strings.TrimPrefix(strings.TrimSpace(splitMessage[1]), "~")
+	return &model.CommandResponse{}
+}
 
-		if channelToJoin, err = a.GetChannelByName(c, targetChannelName, args.TeamId, false); err != nil {
-			return &model.CommandResponse{
-				Text: args.T("api.command_invite.channel.error", map[string]any{
-					"Channel": targetChannelName,
-				}),
-				ResponseType: model.CommandResponseTypeEphemeral,
+func (i *InviteProvider) parseMessage(a *app.App, c request.CTX, args *model.CommandArgs, message string) ([]*model.User, []*model.Channel, *model.CommandResponse) {
+	splitMessage := strings.Split(message, " ")
+
+	targetUsers := make([]*model.User, 0, 1)
+	targetChannels := make([]*model.Channel, 0)
+
+	for j, msg := range splitMessage {
+		if msg == "" {
+			continue
+		}
+
+		if msg[0] == '@' || j == 0 {
+			targetUsername := strings.TrimPrefix(msg, "@")
+			userProfile := i.getUserProfile(a, targetUsername)
+			if userProfile == nil {
+				return targetUsers, targetChannels, &model.CommandResponse{
+					Text:         args.T("api.command_invite.missing_user.app_error"),
+					ResponseType: model.CommandResponseTypeEphemeral,
+				}
 			}
+			targetUsers = append(targetUsers, userProfile)
+		} else {
+			targetChannelName := strings.TrimPrefix(msg, "~")
+			channelToJoin, err := a.GetChannelByName(c, targetChannelName, args.TeamId, false)
+			if err != nil {
+				return targetUsers, targetChannels, &model.CommandResponse{
+					Text: args.T("api.command_invite.channel.error", map[string]any{
+						"Channel": targetChannelName,
+					}),
+					ResponseType: model.CommandResponseTypeEphemeral,
+				}
+			}
+			targetChannels = append(targetChannels, channelToJoin)
 		}
-	} else {
-		channelToJoin, err = a.GetChannel(c, args.ChannelId)
+	}
+
+	if len(targetUsers) == 0 {
+		return targetUsers, targetChannels, &model.CommandResponse{
+			Text:         args.T("api.command_invite.missing_user.app_error"),
+			ResponseType: model.CommandResponseTypeEphemeral,
+		}
+	}
+
+	if len(targetChannels) == 0 {
+		channelToJoin, err := a.GetChannel(c, args.ChannelId)
 		if err != nil {
-			return &model.CommandResponse{
+			return targetUsers, targetChannels, &model.CommandResponse{
 				Text:         args.T("api.command_invite.channel.app_error"),
 				ResponseType: model.CommandResponseTypeEphemeral,
 			}
 		}
+		targetChannels = append(targetChannels, channelToJoin)
+	}
+
+	return targetUsers, targetChannels, nil
+}
+
+func (i *InviteProvider) getUserProfile(a *app.App, username string) *model.User {
+	userProfile, nErr := a.Srv().Store().User().GetByUsername(username)
+	if nErr != nil {
+		return nil
+	}
+
+	if userProfile.DeleteAt != 0 {
+		return nil
 	}
 
+	return userProfile
+}
+
+func (i *InviteProvider) addUserToChannel(a *app.App, c request.CTX, args *model.CommandArgs, userProfile *model.User, channelToJoin *model.Channel) *model.CommandResponse {
```

Comment:
Not a blocker, the checks `HasPermissionToChannel` happens multiple times per channel for the inviter since it runs in  for loop [here](https://github.com/mattermost/mattermost-server/pull/21726/files#diff-a89dd0f0773d551154eba4a48e7fc7100d503018692696cd957038decd42b0fdR56).

I wonder if it's possible to create a function called `addUsersToChannel` which would check permissions for a channel with an inviter, and if everything passes [this line](https://github.com/mattermost/mattermost-server/pull/21726/files#diff-a89dd0f0773d551154eba4a48e7fc7100d503018692696cd957038decd42b0fdR201) can be run in a for loop. What do you think?

Reply:
@plant99 That is a valid concern, especially if, for example, someone is adding 100s of accounts to a single channel using this method. I'll look into it – thanks!